### PR TITLE
[SPARK-16515][SQL]set default record reader and writer for script transformation

### PR DESCRIPTION
**Pull request by adrian-wang**
## What changes were proposed in this pull request?

In `ScriptInputOutputSchema`, we read default `RecordReader` and `RecordWriter` from conf. Since Spark 2.0 has deleted those config keys from hive conf, we have to set default reader/writer class name by ourselves. Otherwise we will get `None` for `LazySimpleSerde`, the data written would not be able to read by script. The test case added worked fine with previous version of Spark, but would fail now.
## How was this patch tested?

added a test case in SQLQuerySuite.

### CODE REVIEW

**Comment by adrian-wang** [comment]
This is strange because I can pass the specific test on my local.

**Comment by adrian-wang** [comment]
I have updated my code and switch to use bash as test case. Hope it will work for Jenkins.

**Comment by jameszhouyi** [comment]
Hi,
Cool ! All of my cases relative to transformation script PASSED after applying this PR . Could Spark guys please review this codes to merge this PR ? Thanks a lots !

Best Regards
Yi

**Comment by chenghao-intel** [comment]
LGTM.

cc @yhuai @liancheng 
This breaks the existed application which using the default delimiter, and we've already verified in TPCx-BB.

**Comment by jameszhouyi** [comment]
Hi Spark guys,
Could you please help to review this PR to merge it in Spark 2.0.0 ? Thanks in advance !

Best Regards,
Yi

**Comment by rxin** [comment]
What do you mean that "Since Spark 2.0 has deleted those config keys from hive conf" ?

**Comment by adrian-wang** [comment]
@rxin  In Spark 2.0, those conf values start with "hive.", which have default value in HiveConf, cannot get the default value now.

**Comment by chenghao-intel** [comment]
HiveConf provides default value `org.apache.hadoop.hive.ql.exec.TextRecordReader`, `org.apache.hadoop.hive.ql.exec.TextRecordWriter` for keys `hive.script.recordreader` and `hive.script.recordwriter` respectively; however, SQLConf doesn't provides those keys, and it means the default values will be null; this causes the backward-incompatibility;

**Comment by rxin** [comment]
Are all script transforms broken? Don't we already have a test case that actually run script transforms?

**Comment by adrian-wang** [comment]
@rxin Only those script transformation cases which use LazySimpleSerde would be affected.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -1329,7 +1329,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
 
         // SPARK-10310: Special cases LazySimpleSerDe
         val recordHandler = if (name == "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe") {
-          Try(conf.getConfString(configKey)).toOption
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Should we just use `getConfString`? That `defaultRecordHandler` method seems unnecessary.

**Comment by rxin** [comment]
@jameszhouyi / adrian-wang / @chenghao-intel 

related to this pull request, we want to have a native implementation for ScriptTransform that does not depend on Hive's serdes. Can you let me know what features are missing from the current native implementation that is not lazysimpleserde? What does lazysimpleserde actually support that the built-in implementation does not?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 1 above)
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: The default value is different for different `key`, you mean to inline the `defaultRecordHandler` function?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -1340,10 +1340,17 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
 
+    def defaultRecordHandler(configKey: String): Option[String] = {
+      Try(conf.getConfString(configKey)).orElse(Try(configKey match {
+        case "hive.script.recordreader" => "org.apache.hadoop.hive.ql.exec.TextRecordReader"
+        case "hive.script.recordwriter" => "org.apache.hadoop.hive.ql.exec.TextRecordWriter"
+      })).toOption
+    }
+
     val (inFormat, inSerdeClass, inSerdeProps, reader) =
       format(inRowFormat, "hive.script.recordreader")
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Can we pass in the default value for the reader/writer? like `format(inRowFormat, "hive.script.recordreader", "org.apache.hadoop.hive.ql.exec.TextRecordReader")` and `format(outRowFormat, "hive.script.recordwriter", "org.apache.hadoop.hive.ql.exec.TextRecordWriter")`. Then, in `def format`, we just use `getConfString(key: String, defaultValue: String)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 3]
```diff
@@ -1306,7 +1306,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
 
     // Decode and input/output format.
     type Format = (Seq[(String, String)], Option[String], Seq[(String, String)], Option[String])
-    def format(fmt: RowFormatContext, configKey: String): Format = fmt match {
+    def format(fmt: RowFormatContext, configKey: String, configValue: String): Format = fmt match {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: The name should show that this value is default value, right?

**Pull request closed without merging**